src/mug.gleam

```diff
@@ -109,13 +177,79 @@ pub type ConnectionOptions {
     /// will also return a timeout, even if this timeout value has not been
     /// reached yet.
     timeout: Int,
+    /// TLS options
+    tls_opts: TlsConnectionOptions,
+  )
+}
+
+pub type TlsConnectionOptions {
+  /// Do not use TLS, use plain TCP
+  NoTls
+  /// Start with a TLS connection
+  UseTls(TlsOptions)
+}
+
+/// Configuration for TLS connections.
+pub type TlsOptions {
+  TlsOptions(verification_method: TlsVerificationMethod)
+}
+
+/// The certificates to use 
+pub type TlsVerificationMethod {
+  /// Uses these CA certificates and regular certificates to verify the server's certificate.
+  ///
+  /// The `use_system_cacerts` option makes mug use the system's CA certificates, which are 
+  /// usually set to a group of competent CAs who sign most of the web's certificates.  
+  /// You may specifiy your own CA certificates with the `cacerts` option, and custom
+  /// certificates and their keys with the `certs_keys` option.
+  ///
+  /// Note that specifying a PEM encoded CA certificate file will result in the system CA
+  /// certificates not being used. This is because of how the underlying [erlang implementation](https://www.erlang.org/doc/apps/ssl/ssl.html#t:client_option/0)
+  /// works. System CA certificates are passed as DER-encoded certificates, and DER encoded
+  /// certs override PEM-encoded ones. Therefore, if you wish to use a PEM encoded CA cert along
+  /// with the system's CA, you should decode the PEM into a DER using (`public_key:pem_decode`)[https://www.erlang.org/doc/apps/public_key/public_key.html#pem-api]
+  /// and use the DerEncodedCaCertificates variant instead.
+  Certificates(
+    use_system_cacerts: Bool,
+    cacerts: Option(CaCertificates),
+    certs_keys: List(CertsKeys),
   )
+  /// Do not verify certificates. While this does allow you to use self-signed certificates.
+  /// It is highly recommended to not skip verification, add a custom CA instead.
+  DangerouslyDisableVerification
+}
+
+/// The CA certificates to use
+pub type CaCertificates {
+  /// Use these der-encoded certificates as CA certificates.
+  DerEncodedCaCertificates(cacerts: List(BitArray))
+  /// Path to a pem-encoded file which contains CA certificates.
+  /// If this option is specified, then the system CA will not be used.
+  PemEncodedCaCertificates(cacertfile: String)
+}
+
+pub type CertsKeys {
+  /// A list of DER-encoded certificates and their corresponding key.
+  DerEncodedCertsKeys(cert: List(BitArray), key: Key)
+  /// Path to a file containing PEM-encoded certificates and their key, with an optional
+  /// password associated with the file containing the key.
+  PemEncodedCertsKeys(
+    certfile: String,
+    keyfile: String,
+    password: Option(BitArray),
+  )
+}
+
+pub type Key {
+  /// A der-encoded key.  
+  /// `alg` is one of 'RSAPrivateKey' | 'DSAPrivateKey' | 'ECPrivateKey' | 'PrivateKeyInfo'.
+  DerEncodedKey(alg: String, key: BitArray)
```

Comment:
```suggestion
  DerEncodedKey(algorithm: String, key: BitArray)
```

If the string has to be one of a known finite variants then this should be a custom type rather than a string.